src/app/ui/editor/moving_cel_state.cpp

```diff
@@ -326,4 +313,54 @@ gfx::Point MovingCelState::intCelOffset() const
                     int(std::round(m_celOffset.y)));
 }
 
+bool MovingCelState::restoreCelStartPosition() const
+{
+  bool modified = false;
+
+  // Here we put back all cels into their original coordinates (so we
+  // can add the undo information from the start position).
+  for (size_t i=0; i<m_celList.size(); ++i) {
+    Cel* cel = m_celList[i];
+    const gfx::RectF& celStart = m_celStarts[i];
+
+    if (cel->layer()->isReference()) {
+      if (cel->boundsF() != celStart) {
+        cel->setBoundsF(celStart);
+        modified = true;
+      }
+    }
+    else {
+      if (cel->bounds() != gfx::Rect(celStart)) {
+        cel->setBounds(gfx::Rect(celStart));
+        modified = true;
+      }
+    }
+  }
+  return modified;
+}
+
+void MovingCelState::onBeforeCommandExecution(CommandExecutionEvent& ev)
+{
+  if (ev.command()->id() == CommandId::Undo()) {
+    restoreCelStartPosition();
+    Doc* document = m_editor->document();
+    // Restore the mask visibility.
+    if (m_maskVisible) {
+      document->setMaskVisible(m_maskVisible);
+      document->generateMaskBoundaries();
+    }
+    m_editor->backToPreviousState();
+    m_editor->releaseMouse();
+  }
+  else if (ev.command()->id() == CommandId::Cancel()) {
+    restoreCelStartPosition();
+    gfx::Point point(0, 0);
+    MouseMessage mouseMsg(MessageType::kMouseUpMessage,
+                          PointerType::Mouse,
+                          MouseButton::kButtonLeft,
+                          KeyModifiers::kKeyNoneModifier,
+                          point);
+    m_editor->getState()->onMouseUp(m_editor, &mouseMsg);
```

Comment:
Why do we need to call onMouseUp() of the current state (which is this MovingCelState)? instead of executing the same code that onMouseUp() does? (which looks pretty similar to the undo case)

I'm not sure what was causing the crash with the previous commit.